### PR TITLE
new concurrency model :single

### PR DESCRIPTION
**Pull request by ph**
This plugin in the current state is not thread safe and we should rely to the :single delegator.

### CODE REVIEW

File: lib/logstash/outputs/s3.rb [hunk 1]
```diff
@@ -175,6 +177,7 @@ def write_on_bucket(file)
         # prepare for write the file
         object = bucket.objects[remote_filename]
         object.write(fileIO,
+                     :upload_id => remote_filename,
```
Comment (ph) [inline lib/logstash/outputs/s3.rb]: AWS, changes the write a bit on patch level.

Comment (andrewvc) [inline lib/logstash/outputs/s3.rb]: Should there be a change in the gemspec to account for this?

**Comment by ph** [comment]
I've let the plugin for a few time and there is threadsafe issue in the current code that prevent it to use `shared` so I've taken the path of least resistance and used `single`. The problem is when elements get deleted from disk but are already in the queue. The page counter increment isnt threadsafe too.

The rewrite I did is threadsafe and uses the `shared` model. (see branch https://github.com/ph/logstash-output-s3/tree/feature/cleaner-code)

File: lib/logstash/outputs/s3.rb [hunk 2]
```diff
@@ -79,6 +79,8 @@ class LogStash::Outputs::S3 < LogStash::Outputs::Base
   config_name "s3"
   default :codec, 'line'
 
+  concurrency :single
```
Comment (jsvd) [inline lib/logstash/outputs/s3.rb]: humm the PR name is "new concurrency model :shared" and the description says the plugin threadsafe, however this sets the model to `:single`? this means that 1 instance is shared among all threads and access to it is globally synchronized.

Comment (ph) [inline lib/logstash/outputs/s3.rb]: yewps, this plugin is in fact in this current state non threadsafe, I've made the plugin run for a few hours to discover that problem.

Comment (ph) [inline lib/logstash/outputs/s3.rb]: I will update the description

**Comment by jsvd** [comment]
LGTM

**Pull request closed without merging**